File: azure-functions-archetype/src/main/resources/archetype-resources/src/test/java/FunctionTest.java
package $package;

import org.junit.Test;

import com.microsoft.azure.functions.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.logging.Logger;

import static org.junit.Assert.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;


/**
 * Unit test for Function class.
 */
public class FunctionTest {
    /**
     * Unit test for hello method.
     */
    @Test
    public void testHttpTriggerJava() throws Exception {
        // Setup
        final HttpRequestMessage<Optional<String>> req = mock(HttpRequestMessage.class);

        final Map<String, String> queryParams = new HashMap<>();
        queryParams.put("name", "Azure");
        doReturn(queryParams).when(req).getQueryParameters();

        final Optional<String> queryBody = Optional.empty();
        doReturn(queryBody).when(req).getBody();


        final HttpResponseMessage.Builder builder = mock(HttpResponseMessage.Builder.class);
        doReturn(builder).when(req).createResponseBuilder(any(HttpStatus.class));
        doReturn(builder).when(builder).body(anyString());

        final HttpResponseMessage res = mock(HttpResponseMessage.class);
        doReturn(res).when(builder).build();

        final ExecutionContext context = mock(ExecutionContext.class);
        doReturn(Logger.getGlobal()).when(context).getLogger();

        // Invoke
        final HttpResponseMessage ret = new Function().HttpTriggerJava(req, context);

        // Verify
        assertSame(res, ret);
    }
}

File: azure-functions-kotlin-archetype/src/main/resources/archetype-resources/src/test/kotlin/FunctionTest.kt
package $package

import com.microsoft.azure.functions.*
import org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito
import org.mockito.Mockito.*
import java.util.Optional
import java.util.logging.Logger
import kotlin.collections.HashMap

/**
 * Unit test for Function class.
 */
class FunctionTest {

    inline fun <reified T : Any> mock() = Mockito.mock(T::class.java)

    private fun testHttpTrigger(httpMethod: HttpMethod) {
        // Setup
        val req = mock<HttpRequestMessage<Optional<String>>>()

        val queryParams = HashMap<String, String>()
        queryParams["name"] = "Azure"
        doReturn(queryParams).`when`<HttpRequestMessage<Optional<String>>>(req).queryParameters

        val queryBody = Optional.empty<String>()
        doReturn(queryBody).`when`<HttpRequestMessage<*>>(req).body
        doReturn(httpMethod).`when`<HttpRequestMessage<*>>(req).httpMethod

        doAnswer { invocation ->
            val status = invocation.arguments[0] as HttpStatus
            HttpResponseMessageMock.HttpResponseMessageBuilderMock().status(status)
        }.`when`<HttpRequestMessage<*>>(req).createResponseBuilder(any(HttpStatus::class.java))

        val context = mock(ExecutionContext::class.java)
        doReturn(Logger.getGlobal()).`when`(context).logger

        // Invoke
        val ret = Function().run(req, context)

        // Verify
        assertEquals(ret.status, HttpStatus.OK)
    }

    /**
     * Unit test for HttpTrigger GET method.
     */
    @Test
    @Throws(Exception::class)
    fun testHttpTriggerGET() {
        testHttpTrigger(HttpMethod.GET)
    }

    /**
     * Unit test for HttpTrigger POST method.
     */
    @Test
    @Throws(Exception::class)
    fun testHttpTriggerPOST() {
        testHttpTrigger(HttpMethod.POST)
    }

}
